Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
+        NSURLConnection* connection = [[[self alloc] initWithRequest:request
+                                                            delegate:[_NSURLConnection_DataDelegate delegateWithQueue:queue handler:handler]
+                                                    startImmediately:NO] autorelease];
+        [connection setDelegateQueue:queue];
+        [connection start];
+    }
 }
 
 /**
  @Status Interoperable
 */
 - (void)dealloc {
-    [_request release];
-    [_protocol release];
-    [_delegate release];
-    [_response release];
+    [_originalRequest release];
     [super dealloc];
 }
 
 /**
  @Status Interoperable
 */
 - (void)start {
-    if (!_didRetain) {
-        [self retain];
-        _didRetain = TRUE;
+    @synchronized(self) {
+        NSURLProtocol* protocol = _protocolForRequest(_originalRequest, self);
+        if (!protocol) {
+            TraceError(TAG, L"[NSURLConnection start] could not create a NSURLProtocol for its NSURLRequest.");
+            return;
+        }
+        _protocol = protocol;
+
+        if (_delegateQueue.get()) {
+            [_delegateQueue addOperation:[NSBlockOperation blockOperationWithBlock:^void() {
+                                [_protocol startLoading];
+                            }]];
+
+        } else if (_scheduledRunLoops.size() == 1) {
+            // TODO #: Cannot currently schedule on more than one runloop
+            for (auto pair : _scheduledRunLoops) {
+                [pair.first performSelector:@selector(startLoading) target:_protocol argument:nil order:0 modes:@[ pair.second.get() ]];
+            }
+
+        } else {
+            // Run on the current run loop in the default mode
+            [[NSRunLoop currentRunLoop] performSelector:@selector(startLoading)
+                                                 target:_protocol
+                                               argument:nil
+                                                  order:0
+                                                  modes:@[ NSDefaultRunLoopMode ]];
+        }
     }
+}
 
-    [_protocol startLoading];
+/**
+ @Status Caveat
+ @Notes  TODO #: Cannot currently schedule on more than one runloop
+                 Callbacks do not currently occur on scheduled runloop
+*/
+- (void)scheduleInRunLoop:(NSRunLoop*)aRunLoop forMode:(NSRunLoopMode)mode {
+    @synchronized(self) {
+        if (_delegateQueue.get()) {
+            [NSException raise:NSInternalInconsistencyException
+                        format:@"A connection cannot be scheduled with both an operation queue and a run loop."];
+        }
+
+        if (_scheduledRunLoops.size() >= 1) {
+            TraceError(TAG, L"NSURLConnection: Scheduling on more than one run loop is currently not supported.");
+            return;
+        }
+
+        StrongId<NSRunLoop> retainedLoop = aRunLoop;
+        StrongId<NSString> retainedMode = mode;
+        _scheduledRunLoops.push_back(std::make_pair(retainedLoop, retainedMode));
```

Comment:
nit: prefer emplace_back; you get to pass the constructor arguments directly, and these three lines flatten to 

```
xxx.emplace_back(aRunLoop, mode);
```

This also avoids the +1/-1/+1/-1 on both runloop and mode.